src/CalcViewModel/Common/DateCalculator.cpp

```diff
@@ -20,13 +20,27 @@ DateCalculationEngine::DateCalculationEngine(_In_ String^ calendarIdentifier)
 // Returns: True if function succeeds to calculate the date else returns False
 bool DateCalculationEngine::AddDuration(_In_ DateTime startDate, _In_ const DateDifference& duration, _Out_ DateTime *endDate)
 {
+    auto currentCalendarSystem = m_calendar->GetCalendarSystem();
+
     try
     {
         m_calendar->SetDateTime(startDate);
 
         if (duration.year != 0)
         {
+            // The Japanese Era system can have multiple year partitions within the same year.
+            // For example, April 30, 2019 is denoted April 30, Heisei 31; May 1, 2019 is denoted as May 1, Reiwa 1.
+            // The Calendar treats Heisei 31 and Reiwa 1 as separate years, which results in some unexpected behaviors where subtracting a year from Reiwa 1 results in a date in Heisei 31.
+            // To provide the expected result across era boundaries, we first convert the Japanese era system to a Gregorian system, do date math, and then convert back to the Japanese era system.
+            // This works because the Japanese era system maintains the same year/month boundaries and durations as the Gregorian system and is only different in display value.
+            if (currentCalendarSystem == CalendarIdentifiers::Japanese)
```

Comment:
Does this special logic need to be specific to adding/subtracing years?  Why not months or days?

Reply:
The issue is really only when we deal with years because Calendar treats Heisei 31 and Reiwa 1 as 2 different years.  When we add a month to a date in the last month of Heisei, it transitions correctly (ie: the same way we transition correctly when we add a month to a date in December on a Gregorian calendar)